teuthology/task/install/__init__.py

```diff
@@ -571,7 +571,7 @@ def task(ctx, config):
         install_overrides = overrides.get('install', {})
         log.debug('INSTALL overrides: %s' % install_overrides)
         teuthology.deep_merge(config, install_overrides.get(project, {}))
-        teuthology.deep_merge(extra_system_packages, install_overrides.get('extra_system_packages', {}))
+        teuthology.deep_merge(extra_system_packages, install_overrides.get('extra_system_packages', []))
```

Comment:
I'm not sure this is going to work in all cases.

`extra_system_packages` is weird -- it can be either a list or a dictionary (with `deb` and `rpm` lists in it).  So, a dictionary is supplied in the config (i.e. `extra_system_packages` variable at this point is a dictionary instead of of empty list, `deep_merge()` would refuse to merge it with an empty list).

Reply:
hmm, this stinks. So if it's a `list` then the packages should be appended to both `deb` and `rpm`? I suppose that's what teuthology should do?

Reply:
@idryomov, do you think this would cover all cases? 
```suggestion
        teuthology.deep_merge(extra_system_packages, install_overrides.get('extra_system_packages', type(extra_system_packages)()))
```
Even if `extra_system_packages` config is `None`, `deep_merge()` function handles the case if `b` is `None`.

Reply:
As to what the override behavior should be, it's a choice that we need to make.  I see at least three options:
1. Require that the form is the same for the original and the override and otherwise the job is malformed (i.e. check that both are lists or both are dictionaries, then merge).
2. Make the override win, both in form and content (i.e. don't attempt to merge anything).
3. Introduce a preprocessing step and transform both the original and the override into a canonical dictionary first, then merge, then hand over to `install` task.

Reply:
> @idryomov, do you think this would cover all cases?

No, I think it would blow up in case the job specifies e.g. a list in the original and a dictionary in the override (i.e. when the defaults don't kick in at all).

Reply:
> > @idryomov, do you think this would cover all cases?
> 
> No, I think it would blow up in case the job specifies e.g. a list in the original and a dictionary in the override (i.e. when the defaults don't kick in at all).

This is actually option 1 from a previous comment.  If folks think it's good enough, I'm all for it ;)

Reply:
Yes, I think that works but perhaps even better would be:

```
overrides_extra_system_packages = install_overrides.get('extra_system_packages')
if overrides_extra_system_packages:
    teuthology.deep_merge(extra_system_packages, overrides_extra_system_packages)
```

It is more verbose but the `type(...)` solution would look baffling to someone unfamiliar with this problem.

Reply:
Oh right, those are 3 great option! I think option 3 adds some additional flexibility whereas option 1 would throw an error and ask the person editing the config to do the same thing teuthology could (make both config into same `type` format). What do you think?

Reply:
works for me